slsim/Sources/QuasarCatalog/quasar_pop.py

```diff
@@ -0,0 +1,343 @@
+        z_value = np.atleast_1d(z_value)
+        denominator = (
+            np.sqrt(np.exp(self.xi * z_value)) + np.sqrt(np.exp(self.xi * self.z_star))
+        ) ** 2
+        result = (
+            -20.90
+            + (5 * np.log10(self.h))
+            - (
+                2.5
+                * np.log10(
+                    np.exp(self.zeta * z_value)
+                    * (1 + np.exp(self.xi * self.z_star))
+                    / denominator
+                )
+            )
+        )
+
+        if np.any(denominator == 0):
+            raise ValueError(
+                "Encountered zero denominator in M_star calculation. Check input values."
+            )
+
+        return result
+
+    def dPhi_dM(self, M, z_value):
+        """Calculates dPhi_dM for a given M and redshift according to Eq (10) in Oguri &
+        Marshall (2010): DOI: 10.1111/j.1365-2966.2010.16639.x.
+
+        :param M: Absolute i-band magnitude.
+        :type M: float or numpy.ndarray
+        :param z_value: Redshift value.
+        :type z_value: float or numpy.ndarray
+        :return: dPhi_dM value.
+        :rtype: float or np.ndarray :unit: mag^-1 Mpc^-3
+        """
+        M = np.atleast_1d(M)
+        z_value = np.atleast_1d(z_value)
+
+        if z_value.shape == ():
+            z_value = np.full_like(M, z_value)
+        if M.shape == ():
+            M = np.full_like(z_value, M)
+
+        alpha_val = np.where(z_value > 3, -2.58, self.alpha)
+        M_star_value = self.M_star(z_value)
+
+        denominator_dphi_dm = (10 ** (0.4 * (alpha_val + 1) * (M - M_star_value))) + (
+            10 ** (0.4 * (self.beta + 1) * (M - M_star_value))
+        )
+
+        # Handle division by zero
+        term1 = np.divide(
+            self.phi_star,
+            denominator_dphi_dm,
+            out=np.full_like(denominator_dphi_dm, np.nan),
+            where=denominator_dphi_dm != 0,
+        )
+
+        return term1
+
+    def convert_magnitude(self, magnitude, z, conversion="apparent_to_absolute"):
+        """Converts between apparent and absolute magnitudes using K-corrections
+        determined in Table 4 of Richards et al. 2006: DOI: 10.1086/503559.
+
+        :param magnitude: Apparent or absolute i-band magnitude.
+        :type magnitude: float or np.ndarray
+        :param z: Redshift.
+        :type z: float or np.ndarray
+        :param conversion: Conversion direction, either 'apparent_to_absolute' or
+            'absolute_to_apparent'.
+        :type conversion: str
+        :return: Converted magnitude.
+        :rtype: float or np.ndarray :unit: mag
+        """
+        # Load the data from the text file as relative path
+        file_path = "tests/TestData/quasar_i_band_K_corrections_Richards_et_al_2006.txt"
+        data = np.loadtxt(file_path)
+
+        # The data is assumed to be in two columns: redshift and K-correction
+        redshifts = data[:, 0]
+        K_corrections = data[:, 1]
+
+        K_corr_interp = interp1d(
```

Comment:
I am not sure how often this function is being called. I'd assume you do it for every quasar (which can be a lot of times). I'd suggest you keep the interpolation as a self. class variable and only load the file and interpolate it one time, and then every other call you only call the interpolated function.

Reply:
In the updated version, the `K_corr_interp` function is now created during the initialization of the `QuasarRate` object and stored as an instance variable. The `convert_magnitude()` function then uses the precomputed interpolation function only once.